ChatSecure/Classes/Controllers/XMPP/OTRXMPPManager.m

```diff
@@ -1035,13 +1036,15 @@ - (void)parseSubscriptionFromRosterItem:(NSXMLElement *)item {
                 buddy.trustLevel = BuddyTrustLevelUntrusted;
                 buddy.displayName = [jid user];
                 buddy.username = [jid bare];
+            } else {
+                buddy = [buddy copy];
             }
             buddy.subscription = [SubscriptionAttributeBridge subscriptionWithString:subscription];
             [buddy setPendingApproval:[ask isEqualToString:@"subscribe"]];
             if ([buddy subscribedFrom]) {
                 [buddy setAskingForApproval:NO];
             }
-            if ([buddy subscribedTo]) {
+            if ([buddy subscribedTo] || [buddy subscribedFrom]) {
                 buddy.trustLevel = BuddyTrustLevelRoster;
```

Comment:
Won't this add _any_ incoming roster items? This delegate callback `- (void)xmppRoster:(XMPPRoster *)sender didReceiveRosterItem:(NSXMLElement *)item;` from XMPPRosterDelegate doesn't distinguish between items that you've added to your roster and items that are incoming presence requests. The `XMPPRosterStorage` version of this callback `- (void)handleRosterItem:(NSXMLElement *)item xmppStream:(XMPPStream *)stream;` filters out non-roster items .

See `isRosterItem:` inside XMPPRoster.m for more details: 

```objc
/**
 * Some server's include in our roster the JID's of user's NOT in our roster.
 * This happens when another user adds us to their roster, and requests permission to receive our presence.
 * 
 * As discussed in RFC 3921, the state of the other user is "None + Pending In",
 * and the server "SHOULD NOT" include these JID's in the roster it sends us.
 * 
 * Nonetheless, some servers do anyway.
 * This method filters out such rogue entries in our roster.
 * 
 * Note that the server will automatically send us the proper presence subscription request,
 * and it will continue to do so everytime we sign in.
 * From the RFC:
 *     the user's server MUST keep a record of the subscription request and deliver the request when the
 *     user next creates an available resource, until the user either approves or denies the request.
 * 
 * So there is absolutely NO reason to process these entries, or include them in the roster's storage.
 * Furthermore, it isn't reliable to depend on these entires being there.
 * The RFC has clearly defined recommendations on the matter, and servers that currently send these rogue items
 * may very likely stop doing so in future versions.
**/
- (BOOL)isRosterItem:(NSXMLElement *)item
{
	NSString *subscription = [item attributeStringValueForName:@"subscription"];
	if ([subscription isEqualToString:@"none"])
	{
		NSString *ask = [item attributeStringValueForName:@"ask"];
		if ([ask isEqualToString:@"subscribe"])
		{
			return YES;
		}
		else
		{
			return NO;
		}
	}
	
	return YES;
}
```

Reply:
Yeah, we can discuss at which point we should set it to TrustLevelRoster, but I think we want the method to handle ALL incoming items, I am not a fan of the isRosterItem method, seems like a hack to me.

Reply:
Hm, yeah.

Whatever we do I think this code should be unified with the RosterStorage code so we don't have two code paths doing different things. Right now rosterItems will flow from the XMPPRoster and end up being processed in two places (or sometimes one, depending on the result of isRosterItem)

```objc
/// inside OTRXMPPManager.m
- (void)xmppRoster:(XMPPRoster *)sender didReceiveRosterItem:(NSXMLElement *)item;
/// inside OTRYapDatabaseRosterStorage.m
- (void)handleRosterItem:(NSXMLElement *)item xmppStream:(XMPPStream *)stream;
```

Coudn't `- (void)parseSubscriptionFromRosterItem:(NSXMLElement *)item` be moved to `OTRYapDatabaseRosterStorage.m`? You could still make an `OTRYapDatabaseRosterStorage: XMPPRosterDelegate` extension and leave the rest of the vCard code inside XMPPManager, since the multicast delegate just lets you add as many delegates as you want.

```objc
// you can have both be delegates
	[self.xmppRoster addDelegate:self delegateQueue:self.workQueue];
	[self.xmppRoster addDelegate:self.xmppRosterStorage delegateQueue:self.workQueue];
```